### ISSUE TITLE
Common.targets doesn't promote WarnAsError to MSBuildTreatWarningsAsErrors

### ISSUE THREAD

**Issue by rainersigwald** [dotnet/msbuild#10871]
Common.targets does respect the "plain property" version of some warning-related settings:

https://github.com/dotnet/msbuild/blob/69b3e7a43fb2fb26812b20807333cf6ca62167f2/src/Tasks/Microsoft.Common.CurrentVersion.targets#L668-L672

but it doesn't do the broader setting `MSBuildTreatWarningsAsErrors` in the same way.

This adds to the confusion about which settings to use.

**Comment by SimaTian** [comment]
We've implemented the intended solution [here](https://github.com/dotnet/msbuild/pull/10942). Then we had to roll it back due to breaking sdk tests [while updating msbuild dependencies.](https://github.com/dotnet/sdk/pull/44977)
The crux of the issue is the fact that some projects started to use non-prefixed version to selectively turn onn TreatWarningsAsErrors - e.g. they have use case like TreatWarningsAsErrors only for this subproject. This is a legitimate use case.
Unfortunately this leaves us in a weird state where our attempt to unify the behavior breaks more people than expected.

Current status:

-  we have a code change ready
-  we have validated that it works
- unfortunately it breaks lot of builds.

There are several possible ways forward:

- We could unify everything except the TreatWarningsAsErrors which is the most commonly used "subproject specific" one. It would bring us somewhat closer to an unified world. I don't like this solution as it creates a weird "exception inside an exception" kind of state, possibly adding another layer of complexity to an already weird behavior.
- We could break people - obviously, we don't want to do that.
- We could create a build check, let it simmer, then break people with a changewave, then break people. This would require lot of babysitting, but maybe it could be a way forward, although not a straightforward one. Also, it still breaks people even if we soften the blow somewhat.
- We can keep the status quo, since it is a minor annoyance for the most part. This might be bad for the nugget issue that initially gave us the motivation to attempt this unification.

@baronfel, what do you think please?

**Comment by baronfel** [comment]
Adding @aortiz-msft for visibility and insight into the impact to NuGet

**Comment by aortiz-msft** [comment]
@nkolev92

**Comment by nkolev92** [comment]
@SimaTian reached out to me offline about this. 

I am comfortable with the 1st proposed path forward in https://github.com/dotnet/msbuild/issues/10871#issuecomment-2529093862. 

MSBuildTreatWarningsAsErrors came after TreatWarningsAsErrors. It will always be a breaking change. I don't have data on how often MSB warnings exist in builds but feels like something that happens quite frequently in large repos. 

I think https://github.com/dotnet/msbuild/issues/10801 is way more important. 
Currently it forces people to duplicate 3 properties. It seems like MSBuild is elevating errors that another component chose not to elevate.
To me naively seems like MSBuildTreatWarningsAsErrors and the other 3 MSBuild prefixed properties should've applied to MSB prefixed warnings only.

**Comment by SimaTian** [comment]
this one got stuck in limbo due to a context switch.
TODO: 
 - reconstruct the timeline and document what was the issue first time around (we had a PR and had to roll it back).
 - then figure out what to do next.

**Comment by voroninp** [comment]
I see the opposite. When I enable `MSBuildTreatWarningsAsErrors` all compiler and analyzer warnings turn into errors. Is it by desing?

**Comment by rainersigwald** [comment]
@voroninp Yes. `MSBuildTreatWarningsAsErrors` means "all warnings of all types from anywhere will be treated as errors by MSBuild".

**Comment by voroninp** [comment]
@rainersigwald , thanks. I thought this setting is only about MsBuild warnings :-)

**Comment by Frulfump** [comment]
> @rainersigwald , thanks. I thought this setting is only about MsBuild warnings :-)

Same lol

**Comment by nkolev92** [comment]
> @voroninp Yes. MSBuildTreatWarningsAsErrors means "all warnings of all types from anywhere will be treated as errors by MSBuild".

I feel like this MSBuildTreatWarningsAsErrors and the likes are broken because of this.

MSBuild doesn't handle the downgrades such as WarningsNotAsErrors or non msbuild warnings, like for example on a warning on a PackageReference such as an audit warning.
If MSBuild was implementing the suppression for all types of warnings, then sure, but it's not.

**Comment by voroninp** [comment]
@nkolev92 @rainersigwald , btw, it's still not documented. Not mentioned in well known MsBuild properties :-)

**Comment by Frulfump** [comment]
> [@nkolev92](https://github.com/nkolev92) [@rainersigwald](https://github.com/rainersigwald) , btw, it's still not documented. Not mentioned in well known MsBuild properties :-)

Yeah that's interesting it's not here https://learn.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-properties?view=visualstudio
But it was added a long time ago here https://github.com/MicrosoftDocs/visualstudio-docs/commit/b9acf94737a831f6afe8658f42b5da24fb3f2c30
So it seems it has since been removed?!

**Comment by KalleOlaviNiemitalo** [comment]
Yes `MSBuildTreatWarningsAsErrors` was removed from the doc page in <https://github.com/MicrosoftDocs/visualstudio-docs/pull/11286>, on the grounds that it doesn't exist in .NET SDK — as it perhaps doesn't, but it still exists in MSBuild.

**Comment by rainersigwald** [comment]
Well that was wrong: MicrosoftDocs/visualstudio-docs#11305 to bring it back.

**Comment by rainersigwald** [comment]
> I feel like this MSBuildTreatWarningsAsErrors and the likes are broken because of this.
> 
> MSBuild doesn't handle the downgrades such as WarningsNotAsErrors or non msbuild warnings, like for example on a warning on a PackageReference such as an audit warning. If MSBuild was implementing the suppression for all types of warnings, then sure, but it's not.

@nkolev92 We have supported `MSBuildWarningsNotAsErrors` for quite a while now, is there some other functionality that we're missing?

**Comment by Frulfump** [comment]
> Well that was wrong: MicrosoftDocs/visualstudio-docs#11305 to bring it back.

The linked PR was auto-closed https://github.com/MicrosoftDocs/visualstudio-docs/pull/11305#issuecomment-4255362310

(Seems like you need a better process for docs changes, some missing SME in the review chain?)

Would be nice if `TreatWarningsAsError`, `MsBuildTreatWarningsAsError` and `/warnaserror`(`-warnaserror`) were documented together somewhere or the docs for each linked to the others. The docs for the last one are also a bit vague about what it means technically and practically.

**Comment by nkolev92** [comment]
I was thinking more about https://github.com/dotnet/msbuild/issues/10801 scenarios. 

I had copilot run through a bunch of scenario and it wasn't able to repro anything other than the 10801 scenarios.

My bad, I made an incorrect assumption that more scenarios were broken.

Seems like some things have been scenarios from https://github.com/dotnet/msbuild/issues/10801 have been fixed as well, but not sure if it's all just yet.
